src/librustc_mir/transform/nrvo.rs

```diff
@@ -0,0 +1,147 @@
+use rustc_index::bit_set::HybridBitSet;
+use rustc_middle::mir::visit::{MutVisitor, PlaceContext};
+use rustc_middle::mir::{self, BasicBlock, Local, Location};
+use rustc_middle::ty::TyCtxt;
+
+use crate::transform::{MirPass, MirSource};
+
+pub struct RenameReturnPlace;
+
+/// Renames the local corresponding to the
+impl<'tcx> MirPass<'tcx> for RenameReturnPlace {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut mir::Body<'tcx>) {
+        /*
+        if tcx.sess.opts.debugging_opts.mir_opt_level == 2 {
+            return;
+        }
+        */
+
+        debug!("Determining if `{:?}` is eligible for NRVO...", src.def_id());
+        let returned_local = match single_local_assigned_to_return_place(body) {
+            Some(l) => l,
+            None => {
+                debug!("MIR was ineligible for NRVO");
+                return;
+            }
+        };
+
+        debug!("MIR was eligible for NRVO, making {:?} the return place", returned_local);
+
+        RenameToReturnPlace { tcx, orig: returned_local }.visit_body(body);
+
+        for block_data in body.basic_blocks_mut() {
+            block_data.statements.retain(|stmt| stmt.kind != mir::StatementKind::Nop);
+        }
```

Comment:
I think you'll also need to do some transformation on the debuginfo of the locals here to merge the return value's debug info with the `returned_local`'s debug info